### PR TITLE
Explicitly set which DQM modules use LuminosityBlock transitions

### PR DESCRIPTION
**Pull request by Dr15Jones**
This change identifies which DQM modules are using LuminosityBlock transitions and explicitly changes the base class they use. The using modules were classified into two categories:

1. Modules creating Lumi based histograms (ie. they call `setLumiFlag`) and
1. Modules which do other work at Lumi boundaries

Separating the two makes it possible to only have those modules which actually have Lumi based histograms to be LuminosityBlockProducers. This decreases the number of data products in the `edm::LuminosityBlock` and allows for more efficient running of the transition.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by schneiml** [comment]
Thanks Chris, this is a cleanup I wanted to have for a while. Were some of the changes done by machine or are all the commits manual work? I am a bit scared by the non-trivial changes, which are always risky when done by hand... would be easier to review if we can separate out the mechanical changes.

As mentioned before, I am not that happy with making things explicitly `one::`, since I don't want them to be `one::` anymore in the mid-term future. Though changing back things should now be a matter of find&replace and not a big deal. 

Also interesting is the final number of modules of the two types. If there are not many, maybe we can remove some features from the DQM infrastructure (and find different solutions). Though, also in the mid-term future, I want *all* DQM output to be produced on a finer-than-runs granularity, which *will* mean throwing a large number of products into the Lumi blocks. If there is any constraint on what can be done using Products (how many different ones to have etc.) we should talk about that now, since my wish was to replace all of `DQMStore` with `edm`-based products and explicit dependencies.

**Comment by Dr15Jones** [comment]
@schneiml all changes were done by 'hand', although I leveraged the compiler and `git grep` where ever I could. The way I proceeded was
1. changed DQMEDAnalyzer to be `one::DQMEDAnalyzer<>` (i.e. dropping `one::DQMLuminosityBlockElements`). This immediately caused all DQM modules which declared the use of `LuminosityBlock` callbacks with `override` to fail to compile.
1. used `git grep setLumiFlag` to find all modules which are generated per Lumi MonitorElements. For all of those modules, I made them inherit explicitly from `one::DQMEDAnalyzer<one::DQMLuminosityBlockElements>`. 
1. systematically went through the list of failing to compile modules to see which ones actually needed LuminosityBlock transitions. Based on what I found I
   1. removed unnecessary calls
   1. switched to using a LuminosityBlock cache if such a conversion was 'easy'
   1. made it a `one::WatchLuminosityBlocks` to return to the behavior it had before the change

So it is possible I missed DQMEDAnalyzers which used LuminosityBlock transitions but did not use the `override` keyword.

**Comment by Dr15Jones** [comment]
> Though, also in the mid-term future, I want all DQM output to be produced on a finer-than-runs granularity, which will mean throwing a large number of products into the Lumi blocks. If there is any constraint on what can be done using Products (how many different ones to have etc.) we should talk about that now, since my wish was to replace all of DQMStore with edm-based products and explicit dependencies.

My take is we should not request the framework to do unnecessary extra work or use extra resources. So if a module does not actually produce something on a LuminosityBlock boundary, we shouldn't require the framework to reserve memory for that product nor call the module on that transition. Right now the DQM system is asking the framework to do a bunch of unnecessary work. In addition, the way it is requesting the unnecessary work is also keeping the framework from being able to efficiently process concurrent LuminosityBlocks.

**Comment by Dr15Jones** [comment]
@schneiml  The goal I want to see is before the end of 2019 I want the prompt RECO job we will run at the Tier 0 for Run 3 to be able to efficiently use concurrent LuminosityBlocks in order to improve our CPU efficiency and increase our event throughput.  That time scale would give us plenty of time to stress test that new configuration before data taking. What is preventing that from being true with our current code is the DQM modules.

The work I was doing here was meant as a stepping stone towards that goal which could be done on that time scale (actually, could be done by the beginning of 2019). One possibility is we use what I did here as an intermediate step towards the larger changes you envision. If we can get things to work using `one` modules then the changes you want to make would not have to be finished before 2019 since we'd already have a working solution we can use before the larger change.

**Comment by Dr15Jones** [comment]
@schneiml how do we move forward?

**Comment by schneiml** [comment]
Nothing immediatly wrong with this PR, I just wanted you to know about my plans. For now, we definitely want the changes in here (even if we might undo some of it at some point, that remains to be seen).

**Comment by Dr15Jones** [comment]
@schneiml Is there anything you need done before you sign this pull request?

**Comment by andrius-k** [comment]
+1

**Comment by Dr15Jones** [comment]
Please review @franzoni @pohsun @tocheng @lpernie @rekovic @thomreis @nsmith-

**Comment by Dr15Jones** [comment]
please review @franzoni @pohsun @tocheng @lpernie @rekovic @thomreis @nsmith-

File: DQM/L1TMonitor/src/L1TRCT.cc [hunk 1]
```diff
@@ -47,20 +48,6 @@ L1TRCT::L1TRCT(const ParameterSet & ps) :
 
   if (verbose_)
     std::cout << "L1TRCT: constructor...." << std::endl;
-
-  outputFile_ =
-      ps.getUntrackedParameter < std::string > ("outputFile", "");
```
Comment (thomreis) [inline DQM/L1TMonitor/src/L1TRCT.cc]: If this parameter is not used anymore it should also be removed from the config files.

File: DQM/L1TMonitor/src/L1TRCT.cc [hunk 2]
```diff
@@ -47,20 +48,6 @@ L1TRCT::L1TRCT(const ParameterSet & ps) :
...
-  if (!outputFile_.empty()) {
-    std::
-	cout << "L1T Monitoring histograms will be saved to " <<
-	outputFile_.c_str() << std::endl;
-  }
-
-  bool disable =
-      ps.getUntrackedParameter < bool > ("disableROOToutput", false);
```
Comment (thomreis) [inline DQM/L1TMonitor/src/L1TRCT.cc]: If this parameter is not used anymore it should also be removed from the config files.

File: DQM/L1TMonitor/src/L1TRPCTF.cc [hunk 3]
```diff
@@ -32,15 +32,6 @@ L1TRPCTF::L1TRPCTF(const ParameterSet& ps)
 
   if(verbose_) cout << "L1TRPCTF: constructor...." << endl;
 
-  outputFile_ = ps.getUntrackedParameter<string>("outputFile", "");
-  if ( !outputFile_.empty() ) {
-    cout << "L1T Monitoring histograms will be saved to " << outputFile_.c_str() << endl;
-  }
-
-  bool disable = ps.getUntrackedParameter<bool>("disableROOToutput", false);
```
Comment (thomreis) [inline DQM/L1TMonitor/src/L1TRPCTF.cc]: If this parameter is not used anymore it should also be removed from the config files.

File: DQM/L1TMonitor/src/L1TRPCTF.cc [hunk 4]
```diff
@@ -32,15 +32,6 @@ L1TRPCTF::L1TRPCTF(const ParameterSet& ps)
 
   if(verbose_) cout << "L1TRPCTF: constructor...." << endl;
 
-  outputFile_ = ps.getUntrackedParameter<string>("outputFile", "");
```
Comment (thomreis) [inline DQM/L1TMonitor/src/L1TRPCTF.cc]: If this parameter is not used anymore it should also be removed from the config files.

File: DQM/L1TMonitor/src/L1TdeRCT.cc [hunk 5]
```diff
@@ -122,20 +123,6 @@ L1TdeRCT::L1TdeRCT(const ParameterSet & ps) :
   if (verbose_)
     std::cout << "L1TdeRCT: constructor...." << std::endl;
 
-  outputFile_ =
-      ps.getUntrackedParameter < std::string > ("outputFile", "");
```
Comment (thomreis) [inline DQM/L1TMonitor/src/L1TdeRCT.cc]: If this parameter is not used anymore it should also be removed from the config files.

File: DQM/L1TMonitor/src/L1TdeRCT.cc [hunk 6]
```diff
@@ -122,20 +123,6 @@ L1TdeRCT::L1TdeRCT(const ParameterSet & ps) :
...
-  if (!outputFile_.empty()) {
-    if(verbose_) std::
-  cout << "L1T Monitoring histograms will be saved to " <<
-  outputFile_.c_str() << std::endl;
-  }
-
-  bool disable =
-      ps.getUntrackedParameter < bool > ("disableROOToutput", false);
```
Comment (thomreis) [inline DQM/L1TMonitor/src/L1TdeRCT.cc]: If this parameter is not used anymore it should also be removed from the config files.

File: DQMOffline/L1Trigger/src/L1TSync_Offline.cc [hunk 7]
```diff
@@ -202,14 +202,6 @@ L1TSync_Offline::L1TSync_Offline(const ParameterSet & pset) :
 
   }
 
-  m_outputFile = pset.getUntrackedParameter < std::string > ("outputFile","");
-
-  if (!m_outputFile.empty()) {
-    std::cout << "L1T Monitoring histograms will be saved to " <<	m_outputFile.c_str() << std::endl;
-  }
-
-  bool disable = pset.getUntrackedParameter < bool > ("disableROOToutput", false);
```
Comment (thomreis) [inline DQMOffline/L1Trigger/src/L1TSync_Offline.cc]: If this parameter is not used anymore it should also be removed from the config files.

File: DQMOffline/L1Trigger/src/L1TSync_Offline.cc [hunk 8]
```diff
@@ -202,14 +202,6 @@ L1TSync_Offline::L1TSync_Offline(const ParameterSet & pset) :
 
   }
 
-  m_outputFile = pset.getUntrackedParameter < std::string > ("outputFile","");
```
Comment (thomreis) [inline DQMOffline/L1Trigger/src/L1TSync_Offline.cc]: If this parameter is not used anymore it should also be removed from the config files.

**Comment by Dr15Jones** [comment]
@thomreis could we do the removal from the config in another pull request? This one has already been waiting for more than a month and is important to get concurrent lumis usable.

**Comment by Dr15Jones** [comment]
@franzoni @lpernie @pohsun @tocheng @fabiocos please review this waiting pull request for ALCA.

**Comment by lpernie** [comment]
+1

**Comment by Dr15Jones** [comment]
@fabiocos what is your plan for integrating this?

**Comment by fabiocos** [comment]
+1

I understand that this update is ok for DQM so far, and @thomreis  L1T cleaning may be done afterwards for simplicity

**Pull request merged**